src/Dotnet.Status.Web/AzureTableInstallationLookup.cs

```diff
@@ -0,0 +1,98 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using GitHubJwt;
+using Microsoft.Dotnet.GitHub.Authentication;
+using Microsoft.DotNet.GitHub.Authentication;
+using Microsoft.Extensions.Options;
+using Octokit;
+
+namespace DotNet.Status.Web
+{
+    public class InMemoryCacheInstallationLookup : IInstallationLookup
+    {
+        private readonly GitHubJwtFactory _jwtFactory;
+        private readonly IOptions<GitHubClientOptions> _options;
+
+        public InMemoryCacheInstallationLookup(
+            GitHubJwtFactory jwtFactory,
+            IOptions<GitHubClientOptions> options)
+        {
+            _jwtFactory = jwtFactory;
+            _options = options;
+        }
+
+        private Dictionary<string, long> _cache = new Dictionary<string,long>();
+        private DateTimeOffset _lastCached = DateTimeOffset.MinValue;
+        private readonly SemaphoreSlim _sem = new SemaphoreSlim(1, 1);
+
+        public async Task<long> GetInstallationId(string repositoryUrl)
+        {
+            var segments = new Uri(repositoryUrl, UriKind.Absolute).Segments;
+            var org = segments[segments.Length - 2].TrimEnd('/');
+
+            if (HasCachedValue(org, out long installation))
+            {
+                return installation;
+            }
+
+            await _sem.WaitAsync();
+            try
+            {
+                if (HasCachedValue(org, out installation))
+                {
+                    return installation;
+                }
+
+                var newCache = new Dictionary<string, long>();
+                string appToken = _jwtFactory.CreateEncodedJwtToken();
+                GitHubAppsClient client = new GitHubAppsClient(
+                    new ApiConnection(
+                        new Connection(_options.Value.ProductHeader)
+                        {
+                            Credentials = new Credentials(appToken, AuthenticationType.Bearer)
+                        }
+                    )
+                );
+
+                foreach (Installation i in await client.GetAllInstallationsForCurrent())
+                {
+                    newCache[org] = i.Id;
+                }
+
+                Interlocked.Exchange(ref _cache, newCache);
+                _lastCached = DateTimeOffset.UtcNow;
+                
+                if (_cache.TryGetValue(org, out installation))
+                {
+                    return installation;
+                }
+                else
+                {
+                    _cache[org] = 0;
+                    return 0;
+                }
+            }
+            finally
+            {
+                _sem.Release();
+            }
+        }
+
+        private bool HasCachedValue(string repositoryUrl, out long installation)
```

Comment:
>HasCachedValue [](start = 21, length = 14)

This will re-fill the cache every time a repo is asked for that doesn't have an installation. This should probably always return true if _lastCached is recent.

Reply:
That was intentional, because if you just installed it somewhere and then came here and it just failed, that would be confusing.  I could try to carry over "we know this didn't exist before, so it doesn't now" when I update the cache.

---
In reply to: [344314874](https://github.com/dotnet/arcade-services/pull/756#discussion_r344314874) [](ancestors = 344314874)

Reply:
The issue is if we ask for 2 different repos back and forth this will spam the github api because we only cache one "0" in the cache.

---
In reply to: [344358231](https://github.com/dotnet/arcade-services/pull/756#discussion_r344358231) [](ancestors = 344358231,344314874)

Reply:
There! I carry over all the previous things we were asked about and call them 0, so we can't ever lose a key, that should solve both the problems. (Also, this file was named wrong, and I fixed it)

---
In reply to: [344359123](https://github.com/dotnet/arcade-services/pull/756#discussion_r344359123) [](ancestors = 344359123,344358231,344314874)